### PR TITLE
Ability to add/update note in VSI

### LINKED ISSUE #93: Allow setting of notes in virtual guest resource
^

### PR DESCRIPTION
**Pull request by ivan-n16**
Capability to add note during provisioning or add/update/remove it afterwards.
Fixes #93

### CODE REVIEW

File: softlayer/resource_softlayer_virtual_guest.go [hunk 1]
```diff
@@ -727,6 +738,11 @@ func resourceSoftLayerVirtualGuestRead(d *schema.ResourceData, meta interface{})
 		}
 	}
 
+	d.Set("notes", nil)
+	if notes := result.Notes; notes != nil {
+		d.Set("notes", *notes)
+	}
+
```
Comment (renier) [inline softlayer/resource_softlayer_virtual_guest.go]:
```go
d.Set("notes", sl.Get(result.Notes, nil))
```

File: softlayer/resource_softlayer_virtual_guest.go [hunk 2]
```diff
@@ -788,9 +804,10 @@ func resourceSoftLayerVirtualGuestUpdate(d *schema.ResourceData, meta interface{
 
 	// Update "hostname" and "domain" fields if present and changed
 	// Those are the only fields, which could be updated
-	if d.HasChange("hostname") || d.HasChange("domain") {
+	if d.HasChange("hostname") || d.HasChange("domain") || d.HasChange("notes") {
 		result.Hostname = sl.String(d.Get("hostname").(string))
 		result.Domain = sl.String(d.Get("domain").(string))
+		result.Notes = sl.String(d.Get("notes").(string))
```
Comment (renier) [inline softlayer/resource_softlayer_virtual_guest.go]: I think we need to refactor this:
```go
if d.HasChange("hostname") {
    result.Hostname = sl.String(d.get("hostname").(string))
}

if d.HasChange("domain") {
    result.Domain = sl.String(d.get("domain").(string))
}

if d.HasChange("notes") {
    result.Notes = sl.String(d.get("notes").(string))
}
```

File: softlayer/resource_softlayer_virtual_guest.go [hunk 3]
```diff
@@ -615,6 +620,12 @@ func resourceSoftLayerVirtualGuestCreate(d *schema.ResourceData, meta interface{
 		return err
 	}
 
+	// Set notes
+	err = setNotes(id, d, meta)
+	if err != nil {
+		return err
+	}
+
```
Comment (renier) [inline softlayer/resource_softlayer_virtual_guest.go:628]: Why can't the notes be set as part of the submission of the rest of the virtual guest data structure (order template above)?

Comment (ivan-n16) [reply softlayer/resource_softlayer_virtual_guest.go:628]: I tried setting it in `getVirtualGuestTemplateFromResourceData()` in the struct 
```
opts := datatypes.Virtual_Guest{
...
Notes: sl.String(d.Get("notes").(string))
}
```

But `notes` did not populate upon provision.

Comment (ivan-n16) [reply softlayer/resource_softlayer_virtual_guest.go:628]: @renier As an example, if I POST the following request with `notes` set,
```
user:apiKey@api.softlayer.com/rest/v3/SoftLayer_Virtual_Guest/createObject
{
  "parameters": [
    {
      "hostname": "new-virtual-server",
      "domain": "test-iv.com",
      "startCpus": "1",
      "maxMemory": "2048",
      "localDiskFlag": "false",
      "hourlyBillingFlag": "true",
      "datacenter": {
        "name": "che01"
      },
      "notes": "Note given during provisioning",
      "operatingSystemReferenceCode": "DEBIAN_7_64"
    }
  ]
}
```

The created virtual server does not have the `notes` field populated.

I found something similar for subnets [here](http://stackoverflow.com/questions/38333724/adding-to-subnets-notes-field-after-ordering) wherein notes cannot be populated at the time of provisioning. It has to be done after provisioning is over.

**Pull request merged**